File: patterns/Creational/Prototype/Go/RealWorldExamples/ComplexPage.go
/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Complex Page Example
|--------------------------------------------------------------------------
| Implement Prototype Design Pattern to create objects without specifying
| the exact class of object that will be created.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype/RealWorldExamples
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-go
|--------------------------------------------------------------------------
*/

package main

import (
    "fmt"
    "time"
)

// Page represents the Prototype.
type Page struct {
    Title    string
    Body     string
    Author   *Author
    Comments []string
    Date     time.Time
}

// NewPage is the constructor for Page.
func NewPage(title, body string, author *Author) *Page {
    page := &Page{
        Title:    title,
        Body:     body,
        Author:   author,
        Comments: []string{},
        Date:     time.Now(),
    }
    author.AddToPage(page)
    return page
}

// AddComment adds a comment to the Page.
func (p *Page) AddComment(comment string) {
    p.Comments = append(p.Comments, comment)
}

// Clone creates a copy of the Page with certain modifications.
func (p *Page) Clone() *Page {
    clone := *p
    clone.Title = "Copy of " + p.Title
    p.Author.AddToPage(&clone)
    clone.Comments = []string{}
    clone.Date = time.Now()
    return &clone
}

// Author represents the author of the Page.
type Author struct {
    Name  string
    Pages []*Page
}

// NewAuthor is the constructor for Author.
func NewAuthor(name string) *Author {
    return &Author{
        Name:  name,
        Pages: []*Page{},
    }
}

// AddToPage adds a Page to the Author's list of pages.
func (a *Author) AddToPage(page *Page) {
    a.Pages = append(a.Pages, page)
}

// Client code demonstrating the Prototype Design Pattern.
func clientCode() {
    author := NewAuthor("John Smith")
    page := NewPage("Tip of the day", "Keep calm and carry on.", author)

    // ...

    page.AddComment("Nice tip, thanks!")

    // ...

    draft := page.Clone()
    fmt.Println("Dump of the clone. Note that the author is now referencing two objects.")
    fmt.Printf("%+v\n", draft)
}

func main() {
    clientCode()
}

File: patterns/Creational/Prototype/JavaScript/RealWorldExamples/ComplexPage.js
/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Complex Page Example
|--------------------------------------------------------------------------
| Implement Prototype Design Pattern to create objects without specifying
| the exact class of object that will be created.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype/RealWorldExamples
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-js
|--------------------------------------------------------------------------
*/

class Page {
    constructor(title, body, author) {
        this.title = title;
        this.body = body;
        this.author = author;
        this.author.addToPage(this);
        this.comments = [];
        this.date = new Date();
    }

    addComment(comment) {
        this.comments.push(comment);
    }

    /**
     * You can control what data you want to carry over to the cloned object.
     *
     * For instance, when a page is cloned:
     * - It gets a new "Copy of ..." title.
     * - The author of the page remains the same. Therefore we leave the
     * reference to the existing object while adding the cloned page to the list
     * of the author's pages.
     * - We don't carry over the comments from the old page.
     * - We also attach a new date object to the page.
     */
    clone() {
        const clone = Object.create(Object.getPrototypeOf(this));
        clone.title = "Copy of " + this.title;
        clone.body = this.body;
        clone.author = this.author;
        clone.author.addToPage(clone);
        clone.comments = [];
        clone.date = new Date();

        return clone;
    }
}

class Author {
    constructor(name) {
        this.name = name;
        this.pages = [];
    }

    addToPage(page) {
        this.pages.push(page);
    }
}

/**
 * The client code.
 */
function clientCode() {
    const author = new Author("John Smith");
    const page = new Page("Tip of the day", "Keep calm and carry on.", author);

    // ...

    page.addComment("Nice tip, thanks!");

    // ...

    const draft = page.clone();
    console.log("Dump of the clone. Note that the author is now referencing two objects.\n\n");
    console.log(draft);
}

clientCode();
